File: src/models/black_scholes.py
import numpy as np
from scipy.stats import norm
from typing import Dict, Tuple

class BlackScholes:
    """
    Black-Scholes option pricing model for European options.
    
    Assumptions:
    - European exercise (only at maturity)
    - No dividends
    - Constant risk-free rate and volatility
    - Log-normal price distribution
    """
    
    def __init__(self, S: float, K: float, T: float, r: float, sigma: float):
        """
        Initialize Black-Scholes model.
        
        Parameters:
        S: Current stock price
        K: Strike price
        T: Time to maturity (years)
        r: Risk-free rate (decimal, e.g., 0.05 for 5%)
        sigma: Volatility (decimal, e.g., 0.20 for 20%)
        """
        self.S = S
        self.K = K
        self.T = T
        self.r = r
        self.sigma = sigma
        
        # Calculate d1 and d2 once
        self.d1 = self._calculate_d1()
        self.d2 = self._calculate_d2()
        
    def _calculate_d1(self) -> float:
        """Calculate d1 parameter."""
        numerator = np.log(self.S / self.K) + (self.r + 0.5 * self.sigma**2) * self.T
        denominator = self.sigma * np.sqrt(self.T)
        return numerator / denominator
    
    def _calculate_d2(self) -> float:
        """Calculate d2 parameter."""
        return self.d1 - self.sigma * np.sqrt(self.T)
    
    def call_price(self) -> float:
        """Calculate European call option price."""
        return (self.S * norm.cdf(self.d1) - 
                self.K * np.exp(-self.r * self.T) * norm.cdf(self.d2))
    
    def put_price(self) -> float:
        """Calculate European put option price."""
        return (self.K * np.exp(-self.r * self.T) * norm.cdf(-self.d2) - 
                self.S * norm.cdf(-self.d1))
    
    def greeks(self) -> Dict[str, Tuple[float, float]]:
        """
        Calculate option Greeks.
        
        Returns:
        Dictionary with Greeks for (call, put):
        - delta: Rate of change of option price with stock price
        - gamma: Rate of change of delta with stock price
        - vega: Sensitivity to volatility
        - theta: Time decay
        - rho: Sensitivity to interest rate
        """
        # Delta
        delta_call = norm.cdf(self.d1)
        delta_put = delta_call - 1
        
        # Gamma (same for call and put)
        gamma = norm.pdf(self.d1) / (self.S * self.sigma * np.sqrt(self.T))
        
        # Vega (same for call and put, in dollars per 1% change in vol)
        vega = self.S * norm.pdf(self.d1) * np.sqrt(self.T) / 100
        
        # Theta (per day)
        theta_call = (-(self.S * norm.pdf(self.d1) * self.sigma) / (2 * np.sqrt(self.T)) -
                     self.r * self.K * np.exp(-self.r * self.T) * norm.cdf(self.d2)) / 365
        theta_put = (-(self.S * norm.pdf(self.d1) * self.sigma) / (2 * np.sqrt(self.T)) +
                    self.r * self.K * np.exp(-self.r * self.T) * norm.cdf(-self.d2)) / 365
        
        # Rho (per 1% change in rate)
        rho_call = self.K * self.T * np.exp(-self.r * self.T) * norm.cdf(self.d2) / 100
        rho_put = -self.K * self.T * np.exp(-self.r * self.T) * norm.cdf(-self.d2) / 100
        
        return {
            'delta': (delta_call, delta_put),
            'gamma': (gamma, gamma),
            'vega': (vega, vega),
            'theta': (theta_call, theta_put),
            'rho': (rho_call, rho_put)
        }
    
    def summary(self) -> None:
        """Print complete option pricing summary."""
        print("=" * 60)
        print("BLACK-SCHOLES OPTION PRICING")
        print("=" * 60)
        print(f"\nInput Parameters:")
        print(f"  Stock Price (S):      ${self.S:.2f}")
        print(f"  Strike Price (K):     ${self.K:.2f}")
        print(f"  Time to Maturity (T): {self.T:.4f} years ({self.T*365:.0f} days)")
        print(f"  Risk-free Rate (r):   {self.r*100:.2f}%")
        print(f"  Volatility (σ):       {self.sigma*100:.2f}%")
        
        print(f"\nIntermediate Values:")
        print(f"  d1: {self.d1:.6f}")
        print(f"  d2: {self.d2:.6f}")
        print(f"  N(d1): {norm.cdf(self.d1):.6f}")
        print(f"  N(d2): {norm.cdf(self.d2):.6f}")
        
        call = self.call_price()
        put = self.put_price()
        
        print(f"\nOption Prices:")
        print(f"  Call Option: ${call:.4f}")
        print(f"  Put Option:  ${put:.4f}")
        
        greeks = self.greeks()
        print(f"\nGreeks:")
        print(f"  {'Greek':<10} {'Call':<12} {'Put':<12}")
        print(f"  {'-'*10} {'-'*12} {'-'*12}")
        for greek_name, (call_val, put_val) in greeks.items():
            print(f"  {greek_name.capitalize():<10} {call_val:>11.6f}  {put_val:>11.6f}")
        
        print("=" * 60)


# Test with known values
if __name__ == "__main__":
    # Example: ATM option, 1 year maturity, 5% rate, 20% vol
    print("\nTest Case 1: At-The-Money Option")
    bs1 = BlackScholes(S=100, K=100, T=1.0, r=0.05, sigma=0.20)
    bs1.summary()
    
    print("\n" * 2)
    print("Test Case 2: Out-of-The-Money Call (ITM Put)")
    bs2 = BlackScholes(S=100, K=110, T=0.5, r=0.05, sigma=0.25)
    bs2.summary()
    
    print("\n" * 2)
    print("Test Case 3: Short-dated Option")
    bs3 = BlackScholes(S=150, K=145, T=30/365, r=0.03, sigma=0.30)
    bs3.summary()

File: java/src/main/java/com/options/models/BlackScholes.java
package com.options.models;

import org.apache.commons.math3.distribution.NormalDistribution;

/**
 * Black-Scholes option pricing model for European options.
 * 
 * Assumptions:
 * - European exercise (only at maturity)
 * - No dividends
 * - Constant risk-free rate and volatility
 * - Log-normal price distribution
 */
public class BlackScholes {
    private final double S;      // Current stock price
    private final double K;      // Strike price
    private final double T;      // Time to maturity (years)
    private final double r;      // Risk-free rate (decimal)
    private final double sigma;  // Volatility (decimal)
    
    // Pre-calculated values for efficiency
    private final double d1;
    private final double d2;
    private final NormalDistribution normal;
    
    /**
     * Initialize Black-Scholes model.
     * 
     * @param S Current stock price
     * @param K Strike price
     * @param T Time to maturity (years)
     * @param r Risk-free rate (decimal, e.g., 0.05 for 5%)
     * @param sigma Volatility (decimal, e.g., 0.20 for 20%)
     */
    public BlackScholes(double S, double K, double T, double r, double sigma) {
        this.S = S;
        this.K = K;
        this.T = T;
        this.r = r;
        this.sigma = sigma;
        this.normal = new NormalDistribution();
        
        // Calculate d1 and d2 once
        this.d1 = calculateD1();
        this.d2 = calculateD2();
    }
    
    /**
     * Calculate d1 parameter.
     * d1 = [ln(S/K) + (r + 0.5*σ²)T] / (σ*√T)
     */
    private double calculateD1() {
        double numerator = Math.log(S / K) + (r + 0.5 * sigma * sigma) * T;
        double denominator = sigma * Math.sqrt(T);
        return numerator / denominator;
    }
    
    /**
     * Calculate d2 parameter.
     * d2 = d1 - σ*√T
     */
    private double calculateD2() {
        return d1 - sigma * Math.sqrt(T);
    }
    
    /**
     * Calculate European call option price.
     * C = S*N(d1) - K*e^(-rT)*N(d2)
     * 
     * @return Call option price
     */
    public double callPrice() {
        return S * normal.cumulativeProbability(d1) - 
               K * Math.exp(-r * T) * normal.cumulativeProbability(d2);
    }
    
    /**
     * Calculate European put option price.
     * P = K*e^(-rT)*N(-d2) - S*N(-d1)
     * 
     * @return Put option price
     */
    public double putPrice() {
        return K * Math.exp(-r * T) * normal.cumulativeProbability(-d2) - 
               S * normal.cumulativeProbability(-d1);
    }
    
    /**
     * Option Greeks for both call and put options.
     * 
     * @return Greeks object containing all Greeks
     */
    public Greeks calculateGreeks() {
        // Delta
        double deltaCall = normal.cumulativeProbability(d1);
        double deltaPut = deltaCall - 1;
        
        // Gamma (same for call and put)
        double gamma = normal.density(d1) / (S * sigma * Math.sqrt(T));
        
        // Vega (same for call and put, in dollars per 1% change in vol)
        double vega = S * normal.density(d1) * Math.sqrt(T) / 100;
        
        // Theta (per day)
        double thetaCall = (-(S * normal.density(d1) * sigma) / (2 * Math.sqrt(T)) -
                            r * K * Math.exp(-r * T) * normal.cumulativeProbability(d2)) / 365;
        double thetaPut = (-(S * normal.density(d1) * sigma) / (2 * Math.sqrt(T)) +
                           r * K * Math.exp(-r * T) * normal.cumulativeProbability(-d2)) / 365;
        
        // Rho (per 1% change in rate)
        double rhoCall = K * T * Math.exp(-r * T) * normal.cumulativeProbability(d2) / 100;
        double rhoPut = -K * T * Math.exp(-r * T) * normal.cumulativeProbability(-d2) / 100;
        
        return new Greeks(deltaCall, deltaPut, gamma, vega, thetaCall, thetaPut, rhoCall, rhoPut);
    }
    
    /**
     * Print complete option pricing summary.
     */
    public void printSummary() {
        System.out.println("==========================================================");
        System.out.println("                BLACK-SCHOLES OPTION PRICING            ");
        System.out.println("==========================================================");
        
        System.out.println("\nInput Parameters:");
        System.out.printf("  Stock Price (S):      $%.2f%n", S);
        System.out.printf("  Strike Price (K):     $%.2f%n", K);
        System.out.printf("  Time to Maturity (T): %.4f years (%.0f days)%n", T, T * 365);
        System.out.printf("  Risk-free Rate (r):   %.2f%%%n", r * 100);
        System.out.printf("  Volatility (σ):       %.2f%%%n", sigma * 100);
        
        System.out.println("\nIntermediate Values:");
        System.out.printf("  d1: %.6f%n", d1);
        System.out.printf("  d2: %.6f%n", d2);
        System.out.printf("  N(d1): %.6f%n", normal.cumulativeProbability(d1));
        System.out.printf("  N(d2): %.6f%n", normal.cumulativeProbability(d2));
        
        double call = callPrice();
        double put = putPrice();
        
        System.out.println("\nOption Prices:");
        System.out.printf("  Call Option: $%.4f%n", call);
        System.out.printf("  Put Option:  $%.4f%n", put);
        
        Greeks greeks = calculateGreeks();
        System.out.println("\nGreeks:");
        System.out.println("  Greek      Call        Put");
        System.out.println("  ---------- ----------- -----------");
        System.out.printf("  Delta      %11.6f  %11.6f%n", greeks.deltaCall, greeks.deltaPut);
        System.out.printf("  Gamma      %11.6f  %11.6f%n", greeks.gamma, greeks.gamma);
        System.out.printf("  Vega       %11.6f  %11.6f%n", greeks.vega, greeks.vega);
        System.out.printf("  Theta      %11.6f  %11.6f%n", greeks.thetaCall, greeks.thetaPut);
        System.out.printf("  Rho        %11.6f  %11.6f%n", greeks.rhoCall, greeks.rhoPut);
        
        System.out.println("==========================================================");
    }
    
    // Getters for the calculated values
    public double getD1() { return d1; }
    public double getD2() { return d2; }
    public double getS() { return S; }
    public double getK() { return K; }
    public double getT() { return T; }
    public double getR() { return r; }
    public double getSigma() { return sigma; }
    
    /**
     * Inner class to hold option Greeks.
     */
    public static class Greeks {
        public final double deltaCall;
        public final double deltaPut;
        public final double gamma;
        public final double vega;
        public final double thetaCall;
        public final double thetaPut;
        public final double rhoCall;
        public final double rhoPut;
        
        public Greeks(double deltaCall, double deltaPut, double gamma, double vega,
                     double thetaCall, double thetaPut, double rhoCall, double rhoPut) {
            this.deltaCall = deltaCall;
            this.deltaPut = deltaPut;
            this.gamma = gamma;
            this.vega = vega;
            this.thetaCall = thetaCall;
            this.thetaPut = thetaPut;
            this.rhoCall = rhoCall;
            this.rhoPut = rhoPut;
        }
    }
}